src/echoclient/EchoClient.php

```diff
@@ -62,51 +65,45 @@ function __construct()
     }
 
     /**
-     * Add an event to echo
+     * Adds a single event to echo
      *
      * @param string $class
      * @param string $source
      * @param array $props
      * @param string|null $userId
+     * @param string|null $timestamp
      * @return bool True if successful, else false
      */
-    public function createEvent($class, $source, array $props = array(), $userId = null)
+    public function createEvent($class, $source, array $props = array(), $userId = null, $timestamp = null)
     {
         $class = ECHO_CLASS_PREFIX . $class;
-        $baseUrl = $this->getBaseUrl();
+        $eventJson = $this->getEventJson($class, $source, $props, $userId, $timestamp);
```

Comment:
Timestamp is already in the API - so adding it here makes sense.

I have slight reservations about it though - because we now stream events into S3 using date based keys. Regardless of what timestamp we specify, the event will end up in S3 with all the other events posted in the same 15 minutes, in a file keyed on now. Not in a file keyed on the provided timestamp.

Maybe this is OK?

Are you planning on using this feature though? Adding it here will inevitably make it more widely used. This story was about creating events in bulk - not about adding the timestamp. If we're not using it - would it be worth holding off adding it here?

Reply:
I had two requirements:
- send bulk events
- make the php client support optional timestamps which is what this does.

Reply:
OK - cool, let's not stop this going down.

I do wonder if this is part of the API we might have issues with over time though.

Another example would be the timeseries tables Chris has implemented in Census.

An event received yesterday would be in yesterdays tables, despite us overriding this timestamp to be last week. When last weeks table is dropped from the view, you might expect the events to disappear - but they won't because they will be in yesterdays table.

If we were adding this into the API itself, I might argue more here, but this is already done, so a moot point maybe.

Reply:
Perhaps we could eventually add some event validation in Echo server to the effect of "if you say the timestamp is X, then we must have received the event by X + 6 hours"